modules/packages/ZMQ.chpl

```diff
@@ -0,0 +1,451 @@
+      if ret == -1 {
+        var errmsg = zmq_strerror(errno):string;
+        writef("Error in ~Context(): %s\n", errmsg);
+      }
+    }
+
+    // socket()
+    proc socket(socktype: int, parSafe: bool = true,
+                linger: int = -1): Socket{
+      var sock = new Socket(this, socktype, parSafe, linger);
+      this.lock = true;
+      this.socks.push_back(sock);
+      this.lock;
+      return sock;
+    }
+
+  } // class Context
+
+  /*
+    ZMQ.Socket
+
+    Design Notes:
+
+    - Do not delete Socket objects; they are deleted when the parent
+      Context object is deleted, however, they can be closed early.
+
+    Questions:
+
+    - Should the parSafe lock be a *counting* lock, so that Socket member
+      functions can call their own locking methods?
+
+  */
+  class Socket {
+
+    // Member Elements
+    const parSafe: bool = true;
+    var lock: sync bool;
+    var socket: c_void_ptr;
+    var context: Context;
+
+    // Constructor
+    proc Socket(ctx: Context, socktype: int, parSafe: bool,
+                linger: int = -1) {
+      this.context = ctx;
+      this.socket = zmq_socket(ctx.ctx, socktype:c_int);
+      if this.socket == nil {
+        var errmsg = zmq_strerror(errno):string;
+        writef("Error in Socket(): %s\n", errmsg);
+      }
+      // Not calling Socket.setsockopt() due to non-recursive binary locking
+      var val = linger:c_int;
+      var ret = zmq_setsockopt(this.socket, ZMQ_LINGER,
+                               c_ptrTo(val):c_void_ptr,
+                               numBytes(c_int):size_t):int;
+      if ret == -1 {
+        var errmsg = zmq_strerror(errno):string;
+        writef("Error in Socket() setting ZMQ_LINGER: %s\n", errmsg);
+      }
+    }
+
+    // Destructor
+    pragma "no doc"
+    proc ~Socket() {
+      if this.socket != nil then
+        this.close();
+    }
+
+    // close
+    proc close() {
+      if this.parSafe then this.lock = true;
+      var ret = zmq_close(this.socket):int;
+      if ret == -1 {
+        var errmsg = zmq_strerror(errno):string;
+        writef("Error in Socket.close(): %s\n", errmsg);
+      }
+      this.socket = c_nil;
+      if this.parSafe then this.lock;
+    }
+
+    // connect
+    proc connect(endpoint: string) {
+      if this.parSafe then this.lock = true;
+      var ret = zmq_connect(this.socket, endpoint.c_str());
+      if this.parSafe then this.lock;
+      if ret == -1 {
+        var errmsg = zmq_strerror(errno):string;
+        writef("Error in Socket.connect(): %s\n", errmsg);
+      }
+    }
+
+    // bind
+    proc bind(endpoint: string) {
+      if this.parSafe then this.lock = true;
+      var ret = zmq_bind(this.socket, endpoint.c_str());
+      if this.parSafe then this.lock;
+      if ret == -1 {
+        var errmsg = zmq_strerror(errno):string;
+        writef("Error in Socket.bind(): %s\n", errmsg);
+      }
+    }
+
+    // setsockopt
+    proc setsockopt(option: int, value: ?T) where isPODType(T) {
```

Comment:
I'd consider also having a setsockopt overload for value is a string. Note that Chapel strings *can* contain NULLs, although you wouldn't want to restrict it to UTF-8, so this would be a "bytes string" whatever that ends up being called (but that's what string is now).